pkg/datasync/directory.go

```diff
@@ -0,0 +1,297 @@
+package datasync
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"sync/atomic"
+	"time"
+
+	dsc3 "github.com/aserto-dev/go-directory/aserto/directory/common/v3"
+	dse3 "github.com/aserto-dev/go-directory/aserto/directory/exporter/v3"
+	"github.com/aserto-dev/go-edge-ds/pkg/bdb"
+	bolt "go.etcd.io/bbolt"
+
+	cuckoo "github.com/panmari/cuckoofilter"
+	"github.com/samber/lo"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+func (s *Sync) syncDirectory(ctx context.Context, conn *grpc.ClientConn) error {
+	runStartTime := time.Now().UTC()
+	s.logger.Info().Str(status, started).Str("mode", s.options.Mode.String()).Msg(syncRun)
+
+	defer func() {
+		close(s.errChan)
+	}()
+
+	// error spew.
+	go func() {
+		for e := range s.errChan {
+			s.logger.Error().Err(e).Msg(syncRun)
+		}
+	}()
+
+	g := new(errgroup.Group)
+
+	g.Go(func() error {
+		err := s.subscriber(ctx)
+		if err != nil {
+			s.logger.Error().Err(err).Str(stage, "subscriber").Msg(syncRun)
+		}
+		return err
+	})
+
+	g.Go(func() error {
+		err := s.producer(ctx, conn)
+		if err != nil {
+			s.logger.Error().Err(err).Str(stage, "producer").Msg(syncRun)
+		}
+		return err
+	})
+
+	if err := g.Wait(); err != nil {
+		s.logger.Debug().Err(err).Msg(syncRun)
+	}
+
+	if Has(s.options.Mode, Diff) {
+		if err := s.diff(ctx); err != nil {
+			s.logger.Error().Err(err).Str(stage, "diff").Msg(syncRun)
+			return err
+		}
+	}
+
+	ts := <-s.tsChan
+	if err := s.setWatermark(ts); err != nil {
+		return err
+	}
+
+	runEndTime := time.Now().UTC()
+	s.logger.Info().Str(status, finished).Str("duration", runEndTime.Sub(runStartTime).String()).Msg(syncRun)
+	return nil
+}
+
+func (s *Sync) producer(ctx context.Context, conn *grpc.ClientConn) error {
+	s.logger.Info().Str(status, started).Msg(syncProducer)
+
+	var recvCtr, objCtr, relCtr int32
+
+	defer func() {
+		s.logger.Debug().Msg("producer closed export channel")
+		close(s.exportChan)
+	}()
+
+	ctx, cancel := context.WithCancel(ctx)
```

Comment:
Why do we need this if we only ever call `cancel()` after the function already existed?

Reply:
Is it to protect from downstream code starting go routines that use the context and leaving them running?